crypto/s390xcap.c: Add guards around the GETAUXVAL checks

levitte opened the pull request:
HWCAP_S390_VX is missing on SUSE Linux Enterprise Server 12 SP1, so we
add a guard that checks the present of that macro.  While we're at it,
we do the same with HWCAP_S390_STFLE, for consistency.

levitte committed:
crypto/s390xcap.c: Add guards around the GETAUXVAL checks

HWCAP_S390_VX is missing on SUSE Linux Enterprise Server 12 SP1, so we
add a guard that checks the present of that macro.  While we're at it,
we do the same with HWCAP_S390_STFLE, for consistency.

p-steuer commented on crypto/s390xcap.c:
```diff
@@ -86,16 +86,20 @@ void OPENSSL_cpuid_setup(void)
     {
```
change
```
#ifdef OSSL_IMPLEMENT_GETAUXVAL
```
to
```
#if defined(OSSL_IMPLEMENT_GETAUXVAL) && defined(HWCAP_STFLE) && defined(HWCAP_S390_VX)
```
so that in case any of the defines is missing, the old signal handler stuff is used.

With your proposed changes, systems that define OSSL_IMPLEMENT_GETAUXVAL ..
- .. but dont define HWCAP_S390_STFLE would disable all HW support.
- .. but dont define HWCAP_S390_VX would run into SIGFPE in case machine has vectory facility but kernel disables it.

levitte replied:
Ah, ok.  Sure

levitte replied:
Fixed

levitte committed:
squash! crypto/s390xcap.c: Add guards around the GETAUXVAL checks

New story:

If one of the macros HWCAP_S390_STFLE or HWCAP_S390_VX is undefined,
we can't support hardware vector calculations.

p-steuer commented on crypto/s390xcap.c:
```diff
@@ -82,7 +82,9 @@ void OPENSSL_cpuid_setup(void)
-#ifdef OSSL_IMPLEMENT_GETAUXVAL
+#if defined(OSSL_IMPLEMENT_GETAUXVAL) \
+    && defined(HWCAP_S390_STFLE) \
+    && defined(HWCAP_S390_VX)
```
i didnt think about it in my last review, but actually the checks for the defines has to be moved up to the conditional definition of OSSL_IMPLEMENT_GETAUXVAL ie: define OSS__IMPLEMENT_GETAUXVAL if and only if all compile time requirements for getauxval code are met (including the 2 defines are available).

levitte replied:
Fixup pushed, please look

p-steuer reviewed (changes requested):
.

levitte committed:
squash! crypto/s390xcap.c: Add guards around the GETAUXVAL checks

p-steuer commented on crypto/s390xcap.c:
```diff
@@ -82,7 +84,7 @@ void OPENSSL_cpuid_setup(void)
     OPENSSL_s390xcap_P.stfle[0] |= S390X_CAPBIT(0);
 
-#ifdef OSSL_IMPLEMENT_GETAUXVAL
+#if defined(OSSL_IMPLEMENT_GETAUXVAL)
```
i dont know why you changed it in this place but left it as "#ifdef" above.

p-steuer commented:
Looks good to me. My systems define HWCAP_S390_VX and HWCAP_S390_STFLE so i didnt test it If it works for you, i approve.

levitte commented:
Merged.

5768804d5ac543b2c899bcbb96cb6f571cef9eaa crypto/s390xcap.c: Add guards around the GETAUXVAL checks
